### PR TITLE
feat(ics20): implement on receive packet

### PR DESCRIPTION
**Pull request by gjermundgaraba**

### CODE REVIEW

File: src/utils/ICS20Lib.sol
Comment (gjermundgaraba) [inline src/utils/ICS20Lib.sol:78]: These were changed to match the order they come in from ibc-go. This is a limitation of the ICS20Lib: it only supports JSON in a specific order. Parsing the JSON in any order is possible, but might be more gas-intensive. Should we create an issue for that? There is a TODO below on this.

File: src/ICS26Router.sol
Comment (srdtrk) [reply src/ICS26Router.sol:133]: Yes, I'd rather have this removed as we want to value simplicity above everything else. Putting a restriction on the user is not problematic (since frontends should take care of it). Also, even if the user sends a bad packet, it'd just timeout.

File: src/ICS26Router.sol
Comment (srdtrk) [reply src/ICS26Router.sol:127]: Let's remove this for now since it doesn't provide much more utility other than debugging but adds lines of code that require reasoning and linter rules

File: script/E2ETestDeploy.s.sol
Comment (srdtrk) [reply script/E2ETestDeploy.s.sol:28]: Yes, in scripts and tests, lets do solhint ignore for custom errors.

File: e2e/interchaintestv8/ibc_eureka_test.go
Comment (srdtrk) [reply e2e/interchaintestv8/ibc_eureka_test.go:509]: Hmm, we probably should remove it since it is not needed

File: e2e/interchaintestv8/e2esuite/grpc_query.go
Comment (srdtrk) [reply e2e/interchaintestv8/e2esuite/grpc_query.go:45]: You could use `GRPCQuery` function below I think

**Review by srdtrk — COMMENTED** [review]
Amazing work as always

File: src/utils/ICS20Lib.sol
Comment (srdtrk) [reply src/utils/ICS20Lib.sol:78]: Yes, let's create an issue. And please link the issues to the todo items (as numbers) if there is an associated issue.

File: src/ICS20Transfer.sol
Comment (srdtrk) [reply src/ICS20Transfer.sol]: I did this

File: src/ICS26Router.sol
Comment (srdtrk) [reply src/ICS26Router.sol]: I've reverted the nanosecond changes, if this becomes an issue in the future, I'd consider only using nanoseconds (instead of seconds). But supporting both is too much unneeded complexity and additional security considerations.

File: src/ICS26Router.sol
Comment (srdtrk) [reply src/ICS26Router.sol]: I removed these

File: e2e/interchaintestv8/ibc_eureka_test.go
Comment (srdtrk) [reply e2e/interchaintestv8/ibc_eureka_test.go]: I've removed it

**Review by srdtrk — APPROVED** [review]
I'm approving after the pushes I've made. Great work. I'll let you respond to my review before merging this :)

File: src/ICS26Router.sol
Comment (gjermundgaraba) [reply src/ICS26Router.sol]: Alright, makes sense. After having dealt with this for a bit, I feel like the devUX is pretty clunky when there are different timestamps. Unless we can change ibc-go, I would suggest in the future that we seriously consider just using nanoseconds everywhere. But for now, this works :)

File: src/ICS20Transfer.sol
Comment (gjermundgaraba) [reply src/ICS20Transfer.sol]: Great, thanks! :)

File: e2e/interchaintestv8/e2esuite/grpc_query.go
Comment (gjermundgaraba) [reply e2e/interchaintestv8/e2esuite/grpc_query.go:45]: I tried, but the method/path came out wrong, for some reason. 🤷

**Pull request merged**
